src/webview/html/messageAsHtml.js

```diff
@@ -72,13 +78,90 @@ $!${messageReactionListAsHtml(reactions, ownUser.user_id, allImageEmojiById)}
 `;
 };
 
-const widgetBody = (message: Message | Outbox) => template`
+/**
+ * Render the body of a message that has submessages.
+ *
+ * Must not be called on a message without any submessages.
+ */
+const widgetBody = (message: Message, ownUserId: UserId) => {
+  invariant(
+    message.submessages !== undefined && message.submessages.length > 0,
+    'should have submessages',
+  );
+
+  const sortedSubmessages = message.submessages.concat();
+  sortedSubmessages.sort((m1, m2) => m1.id - m2.id);
+
+  const widgetSubmessages: $ReadOnlyArray<{
+    sender_id: number,
+    content: SubmessageData,
+  }> = sortedSubmessages
+    .filter(submessage => submessage.msg_type === 'widget')
+    .map(submessage => ({
+      sender_id: submessage.sender_id,
+      content: JSON.parse(submessage.content),
+    }));
+
+  const pollWidget = widgetSubmessages[0];
+  if (!pollWidget || !pollWidget.content || pollWidget.content.widget_type !== 'poll') {
+    return template`
 $!${message.content}
 <div class="special-message"
  ><p>Interactive message</p
  ><p>To use, open on web or desktop</p
 ></div>
 `;
+  }
+
+  if (pollWidget.content.extra_data == null) {
+    // We don't expect this to happen in general, but there are some malformed
+    // messages lying around that will trigger this.
+    return template`$!${message.content}`;
+  }
+
+  const pollData = new PollData({
+    message_sender_id: message.sender_id,
+    current_user_id: ownUserId,
+    is_my_poll: message.sender_id === ownUserId,
+    question:
+      pollWidget.content.extra_data.question == null ? pollWidget.content.extra_data.question : '',
```

Comment:
nit: Can equivalently say `pollWidget.content.extra_data.question ?? ''`, and that makes it a bit more obvious what's going on.

Reply:
That sounds lovely, but when I try it, flow reports:

```
Cannot get pollWidget.content.extra_data.question because property question is missing in property
extra_data of unknown type [1]. [incompatible-use]

 [1] 116│   if (pollWidget.content.extra_data == null) {
        :
     123│     message_sender_id: message.sender_id,
     124│     current_user_id: ownUserId,
     125│     is_my_poll: message.sender_id === ownUserId,
     126│     question: pollWidget.content.extra_data.question ?? '',
```

Reply:
(I actually had the conditional backwards in the code I pushed, but even with the conditional correct, it passes flow, but doesn't with the `??`)

Reply:
FTR this is now fixed by the cast, with a fixme, to `WidgetData`. The error is correct -- this logic is relying on assumptions that aren't in the types, because the structure of these submessages arrays is kind of quirky and can't be expressed in the type system.

(We never did figure out why there wasn't an error in the `? :` case.)